rs/orchestrator/src/ssh_access_manager.rs

```diff
@@ -167,6 +176,512 @@ impl SshAccessManager {
+        let ssh_assigned_readonly_access =
+            test.assigned_readonly.to_ssh_access(ASSIGNED_READONLY_KEY);
+        let ssh_unassigned_readonly_access = test
+            .unassigned_readonly
+            .to_ssh_access(UNASSIGNED_READONLY_KEY);
+        let ssh_backup_access = test.backup.to_ssh_access(BACKUP_KEY);
+        let ssh_recovery_access = test.recovery.to_ssh_access(RECOVERY_KEY);
+
+        let registry_data = Arc::new(ProtoRegistryDataProvider::new());
+        let registry_client =
+            Arc::new(FakeRegistryClient::new(Arc::clone(&registry_data) as Arc<_>));
+
+        if ssh_assigned_readonly_access.is_some() || ssh_backup_access.is_some() {
+            let subnet_record = SubnetRecord {
+                ssh_readonly_access: ssh_assigned_readonly_access.clone().unwrap_or_default(),
+                ssh_backup_access: ssh_backup_access.unwrap_or_default(),
+                membership: vec![ASSIGNED_NODE.get().to_vec()],
+                ..test_subnet_record()
+            };
+            registry_data
+                .add(
+                    &make_subnet_record_key(SUBNET_ID),
+                    REGISTRY_VERSION,
+                    Some(subnet_record),
+                )
+                .expect("Failed to add subnet record.");
+        }
+
+        if let Some(ssh_readonly_access) = ssh_unassigned_readonly_access {
+            registry_data
+                .add(
+                    &make_unassigned_nodes_config_record_key(),
+                    REGISTRY_VERSION,
+                    Some(UnassignedNodesConfigRecord {
+                        ssh_readonly_access,
+                        replica_version: "some_version".into(),
+                    }),
+                )
+                .expect("Failed to add unassigned nodes record.");
+        }
+
+        if let Some(ssh_node_state_write_access) = ssh_recovery_access {
+            for node_id in [ASSIGNED_NODE, API_BOUNDARY_NODE, UNASSIGNED_NODE] {
+                registry_data
+                    .add(
+                        &make_node_record_key(node_id),
+                        REGISTRY_VERSION,
+                        Some(NodeRecord {
+                            ssh_node_state_write_access: ssh_node_state_write_access.clone(),
+                            ..Default::default()
+                        }),
+                    )
+                    .expect("Failed to add node record.");
+            }
+        }
+
+        registry_data
+            .add(
+                &make_api_boundary_node_record_key(API_BOUNDARY_NODE),
+                REGISTRY_VERSION,
+                Some(ApiBoundaryNodeRecord {
+                    version: "some_version".into(),
+                }),
+            )
+            .expect("Failed to add boundary node record.");
+
+        registry_client.reload();
+        registry_client
+    }
+
+    fn setup_ssh_access_manager(
+        node_id: NodeId,
+        test: TestCase,
```

Comment:
```suggestion
        test: &TestCase,
```
I think borrowing is enough (in `setup_registry` too). We could even remove `#[derive(Clone)]` in `TestCase` and `RegistryEntry` in that case.

You could also borrow `logger` and `clone` it twice inside here such that we do not have to clone it at caller sites.

This would turn all `test.clone(), log.clone()` into `&test, &log`.

Reply:
Done: https://github.com/dfinity/ic/pull/7265/commits/275be36262f53016273ee5456ce9f596086b2073